Refactor(#70): 부분 수정 AI 답변을 개선한다

kimyu0218 opened the pull request:
<!-- 🔥 다음 양식으로 제목을 작성해주세요 : 한 일의 type(#issue number): 작업 내용 -->
<!-- ex) Feature(#133): canvas 구현~ -->
<!-- "여기에 작성하세요" 는 지우고 작성하세요 🙏🏻 -->

resolved #70

## 작업 개요
<!-- 작업에 대한 설명을 간단하게 작성해주세요. -->
[이슈](https://github.com/kopilot2024/kopilot/issues/70)에 작성한 문제 해결 및 개선

## 작업 사항
<!-- 작업에 대한 설명을 코드와 관련하여 남겨주세요. -->
- 부분 수정 옵션별 최소 글자수 설정
- 자체 alert 창 제작
- 플로팅 버튼에 대한 툴팁 제작

## 고민한 점들(필수 X)
<!-- 작업을 진행하면서 고민했던 점들을 추가해주세요 -->

제 기준으로 최소 글자수를 작성하긴 했는데 사용자가 불편함을 느낄 수 있는 부분이라 사용해보면서 글자수 타협하면 좋을 것 같아요!!

```ts
export const Command: Record<
  CommandValue,
  { label: CommandLabel; length: number }
> = {
  LONG_DESCRIPTION: { label: '길게 풀어서 작성', length: 20 },
  SHORT_DESCRIPTION: { label: '간결하게 작성', length: 40 },
  SUBTITLE: { label: '소제목 작성', length: 200 },
  SYNONYM: { label: '유의어 대체', length: 1 },
  DIRECT_COMMAND: { label: 'AI에게 직접 명령', length: 0 },
};
```

## 스크린샷(필수 X)
<!-- 작업을 파악하는 데 도움이 되는 스크린샷을 추

https://github.com/user-attachments/assets/ef352c4a-c412-4f98-b95e-ca9d76593482

가해주세요 -->

kimyu0218 committed:
refactor: 글자수 제한 및 프롬프트 변경 [#70]

- 유의어 대체를 제외한 부분 수정 옵션의 경우, 글자수가 적으면 이상한
  답변을 함
- 각 옵션별로 최소 글자수를 설정하여 의미 없는 답변을 받지 않도록 제어
- 유의어 대체는 유의어가 없는 경우에도 반영 버튼이 뜨는 경우가 있음
- 유의어가 없는 경우에는 없다는 표시 제공

kimyu0218 committed:
refactor: alert 팝업 제작 및 팝업 리팩토링 [#70]

- alert 창을 위한 팝업 제작
- 공통 부분 상위 클래스에 정의하도록 수정

kimyu0218 committed:
refactor: data-length에 최소 글자수 저장 [#70]

최소 글자수를 확인하기 위해 data-length에 각 옵션별 최소 글자수 저장

kimyu0218 committed:
docs: alert 팝업을 위한 파일 추가 정의 및 툴팁 제작

- 기존 alert 창 모두 커스텀한 alert 팝업으로 대체
- 플로팅 버튼을 위한 툴팁 추가

ttobe commented on views/javascript/components/basePopup.js:
outputPopup에 있는 내용이 다 위로 올라왔네요 좋은 것 같습니다!

ttobe commented on views/javascript/components/editorBox.js:
```diff
@@ -17,19 +21,26 @@ export class EditorBox extends BaseComponent {
+  show(text, command, length, label) {
+    if (text.length < length && command !== DIRECT_COMMAND) {
+      this.#alertPopup.pop(
+        `"<span>${label}</span>"은(는)${SPACE}<span>${length}자 이상</span>${SPACE}작성해주세요!`,
```
보통 `"${label}"` 이런식으로 하지 않나요? `span`을 `""`로 감싸는 이유가 있나요?

kimyu0218 replied:
`<span>` 태그로 색상 강조 표시 하고 있어요!
이 부분도 이해하기 쉽게 `span class='e'로 바꿔서 `index.css`에서 강조할 때 다 이 태그 사용하면 색상이 칠해지도록 진행하려고 합니다
![image](https://github.com/user-attachments/assets/502e5ba0-fe40-4ce3-954c-ba695cb13dc7)

ttobe commented on views/javascript/components/editorBox.js:
```diff
@@ -102,16 +115,22 @@ export class EditorBox extends BaseComponent {
       'partial modification error',
     );
-    const data = await response.json();
+    const { result } = await response.json();
```
오 `data.result`로 사용하다가 이거 좋은 것 같네요

kimyu0218 replied:
오늘 코드 보다가 진혁님께서 이렇게 사용하신 것 같아서 바꾼 건데요...? ㅋㅋㅋㅋ
승연님 코드였나,,,

tmddus2 replied:
전 아닙니다..ㅋㅋㅋ

ttobe commented:
사용하시는 코드 정리를 되게 많이 하셨네요 고생하셨습니다!

kimyu0218 committed:
remove: 불필요한 content 영역 삭제

kimyu0218 committed:
refactor: 피드백 기준 백엔드에서 관리

- 피드백 기준을 백엔드에서 타입과 상수로 정의하여 관리
- 모델과 템플릿 엔진을 활용하여 데이터를 뿌리도록 수정

kimyu0218 committed:
feat: 템플릿 엔진 JSON 처리를 위한 helper 등록

kimyu0218 committed:
refactor: 공통 로직 제거

- 템플릿 엔진을 이용하여 공통된 코드 제거
- 데이터를 전달하여 버튼 이름 등 내용이 달라지도록 구현

kimyu0218 committed:
docs: CSS 변수 선언

색상, 사이즈, z-index 등 변수로 관리하여 직관적으로 수정

kimyu0218 committed:
refactor: 부분수정 속성 모두 백엔드에서 관리 [#70]

- 부분 수정의 경우 옵션 별로 대체 옵션이 다름
- 소제목은 clova 결과값이 선택 영역 앞, 나머지는 선택 영역의 값을 모두
  대체해야 함
- length처럼 data-position으로 넘겨 백엔드 한 곳에서 관리하도록 리팩토링

kimyu0218 committed:
docs: 자동 완성 색상 수정

글의 흐름을 방해하지 않는 색상으로 변경

kimyu0218 committed:
fix: 글쓰기 툴바가 보이면 textarea가 잘 클릭되지 않는 문제 해결

- 글쓰기 툴바가 띄워지면 textarea를 숨김 처리하여 툴바 바깥 영역을
  클릭해도 textarea를 수정하는 게 쉽지 않았음
- 이를 해결하기 위해 visibility 대신 opacity를 조작하여 해결

kimyu0218 committed:
docs: UI 깨지는 문제 해결 및 반응형 일부 구현

kimyu0218 committed:
docs: textarea-output 폰트 크기 동일하게 설정

kimyu0218 committed:
docs: 폰트 크기 및 단순 에러 수정

kimyu0218 committed:
fix & refactor: 부분수정시 글자 수 업데이트 안되는 문제 해결

- input은 사용자 입력만을 감지하기 때문에 소스코드가 변화시킨 입력을
  감지하지 못함
- Object.defineProperty를 통해 textarea의 value 속성을 재저의하고
  value가 설정될 때마다 input 이벤트를 dispatch 하도록 구현
